protocol_test.go

```diff
@@ -63,3 +66,113 @@ func BenchmarkCanonicalizeContentType(b *testing.B) {
 		b.ReportAllocs()
 	})
 }
+
+func TestProtocolFromRequest(t *testing.T) {
+	t.Parallel()
+	tests := []struct {
+		name        string
+		contentType string
+		method      string
+		want        string
+		valid       bool
+	}{{
+		name:        "connectUnary",
+		contentType: "application/json",
+		method:      http.MethodPost,
+		want:        ProtocolConnect,
+		valid:       true,
+	}, {
+		name:        "connectStreaming",
+		contentType: "application/connec+json",
+		method:      http.MethodPost,
+		want:        ProtocolConnect,
+		valid:       true,
+	}, {
+		name:        "grpcWeb",
+		contentType: "application/grpc-web",
+		method:      http.MethodPost,
+		want:        ProtocolGRPCWeb,
+		valid:       true,
+	}, {
+		name:        "grpc",
+		contentType: "application/grpc",
+		method:      http.MethodPost,
+		want:        ProtocolGRPC,
+		valid:       true,
+	}, {
+		name:        "connectGet",
+		contentType: "application/connec+json",
+		method:      http.MethodGet,
+		want:        ProtocolConnect,
+		valid:       true,
+	}, {
+		name:        "grpcWebGet",
+		contentType: "application/grpc-web",
+		method:      http.MethodGet,
+		want:        ProtocolConnect,
+		valid:       true,
+	}, {
+		name:        "grpcGet",
+		contentType: "application/grpc+json",
+		method:      http.MethodGet,
+		want:        ProtocolConnect,
+		valid:       true,
```

Comment:
This is an odd result and maybe should be revisited to defaulting to `unknown`. We allow any `application/` prefix to fall-through to be classified as a connect request.

Reply:
I don't think that's right. I've got an issue in gRPC with some spec issues/clarifications (https://github.com/grpc/grpc/issues/36767), and the discussion there has brought up possible experimental support for other verbs, like GET, in gRPC.

If a content-type is set, we should likely use it to ascertain the protocol and not look at the HTTP method. If there is no content-type, then we'd see if it's GET and if so also check for the presence of required `message` and `encoding` query params. Otherwise, it's unknown. How does that sound?

Reply:
Sounds good! Now check for query parameters for GET requests. Removed `Content-Type` from all GET requests as that should not be set for any valid request. Still enforce http method for protocol checks as I do think the spec currently requires it, so these are now otherwise unknown.